### DISCUSSION TITLE
Docker not updating code.

### DISCUSSION THREAD

**Discussion by barchakuz** [camicroscope/caMicroscope#825] [Q&A] [answered]
I have followed all instructions(clone repos, built compose docker image),created container in docker as well and launched site as shown in the pictures below. now i have made fews updates in code of locally cloned repo, despite container running smoothly, it doesn't show update.

**Docker Container** 
![const2](https://github.com/camicroscope/caMicroscope/assets/66128370/e04b2216-863e-47e7-ad2c-2ffc0a27af8e)
![cont](https://github.com/camicroscope/caMicroscope/assets/66128370/294aa6f5-169b-4537-addf-8f0a9b389be1)
![tt](https://github.com/camicroscope/caMicroscope/assets/66128370/81489c62-9b59-433d-a67a-352a04feba85)
**Note : I have added some text for and experiment and saved the file but still seeing now change.**
![change](https://github.com/camicroscope/caMicroscope/assets/66128370/296521ba-d9df-4f23-9a06-527e4224c415)
![chng](https://github.com/camicroscope/caMicroscope/assets/66128370/579a1956-17b7-4905-843c-5927f9a05fca)

**Comment by birm** [comment]
can you include the docker compose file you're running from?

**Reply by barchakuz** [reply]
i can't find "docker-compose.yml" file in all three clone repos.Although i have built with "docker-compose -f caMicroscope.yml build" & "docker-compose -f caMicroscope.yml up" as well.

**Reply by birm** [reply]
caMicroscope.yml (or develop.yml if you used that instead) is the file I'm looking to read!

**Reply by barchakuz** [reply]
here is **develop.yml** from Distro repo

  
version: '2'

services:
  mongo:
    image: mongo:4.2-bionic
    container_name: ca-mongo
    volumes:
      - ./db:/data/db
  back:
    build:
      context: "https://github.com/camicroscope/caracal.git#develop"
      args:
        viewer: "develop"
        fork: "camicroscope"
    depends_on:
      - "mongo"
    ports:
      - "4010:4010"
    container_name: ca-back
    volumes:
      - ./config/login.html:/src/static/login.html
      - ./jwt_keys/:/src/keys/
      - ./config/routes.json:/src/routes.json
      - ./config/contentSecurityPolicy.json:/src/contentSecurityPolicy.json
    environment:
      JWK_URL: "https://www.googleapis.com/oauth2/v3/certs"
      IIP_PATH: "http://ca-iip:8080/fcgi-bin/iipsrv.fcgi"
      MONGO_URI: "mongodb://ca-mongo"
      DISABLE_SEC: "true"
      ALLOW_PUBLIC: "true"
  iip:
    build: "https://github.com/camicroscope/iipImage.git#develop"
    container_name: ca-iip
    restart: unless-stopped
    volumes:
      - ./images/:/images/
  loader:
    build: "https://github.com/camicroscope/SlideLoader.git#develop"
    container_name: ca-load
    volumes:
      - ./images/:/images/
      - ./cloud-upload-apis/:/cloud-upload-apis/
    environment:
      DICOM_PORT: "11112"
      DICOM_UI_PORT: "8042"
  dicomsrv:
    build: "https://github.com/camicroscope/dicomsrv.git"
    container_name: ca-dicomsrv
    restart: unless-stopped
    stdin_open: true
    tty: true
    ports:
      - "8042:8042"
      - "11112:11112"
    volumes:
      - ./jwt_keys/:/root/keys/
      - ./images/:/images/
      - ./config/OrthancConfiguration.json:/root/src/Configuration.json
    environment:
      DICOM_PORT: "11112"
      DICOM_UI_PORT: "8042"
      CARACAL_BACK_HOST_PORT: "ca-back:4010"

**Reply by birm** [reply]
if I'm understanding properly, you want to see the changes before committing/pushing. In that case see this section of readme https://github.com/camicroscope/camicroscope?tab=readme-ov-file#fast-local-changes

Also worth noting that you should use docker-compose -f (the file you made changes to)

Hope this helps!

**Reply by We13b-MD** [reply]
@barchakuz @birm how do i set the build to build local changes instead of hosting git versions in my develop.yml file am a bit confused

**Reply by birm** [reply, marked as answer]
By default, ca-back clones a version of camicroscope's static files into itself. The information from above link should let you replace that with whatever you want (your version of caMicroscope) with a docker volume.

The way the guide is written, it expects distro and caMicroscope in the same directory, then you can add that line to develop.yml.

Hope shit helps! If not, we should make the instructions clearer.

**Reply by We13b-MD** [reply]
Thanks birm noted will.let you know when it's working and if there are further improvements

**Reply by We13b-MD** [reply]
Just tested it everything working fine I can now contribute and reflect my additions thank you birm you are awesome 👍 you are the best

**Comment by SanjanaSogimatt** [comment]
Hey @We13b-MD , I wanted some help in running the distro
![error-1](https://github.com/camicroscope/caMicroscope/assets/87171143/6a7661c1-1179-4cc2-a351-d4be65c0a33e)
#Error 1
![error-2](https://github.com/camicroscope/caMicroscope/assets/87171143/acab666e-642a-4b35-b742-7b4ca1860302)
Error 2
I'm getting this even after installing gunicorn and openslide library. And I have set the environment path too
@We13b-MD it would be great if you could help me with this

**Reply by gbengaoluwadahunsi** [reply]
Hello @barchakuz , the issue there is your ca-load, try to run this

1. docker-compose -f develop.yml build loader --no-cache
2. docker-compose -f caMicroscope.yml build
3. docker-compose -f caMicroscope.yml up

it will fix the ca-loader error and run the Distro seamlessly

**Reply by SanjanaSogimatt** [reply]
Hey, @gbengaoluwadahunsi , @We13b-MD the above commands are not helping even after I installed openslide-python on my machine

**Reply by SanjanaSogimatt** [reply]
Hey @gbengaoluwadahunsi @birm @barchakuz @We13b-MD can anyone help me with the above error

**Reply by Seun-Ayela** [reply]
Yes that's exactly what I'm facing at the moment

**Reply by SanjanaSogimatt** [reply]
@Seun-Ayela make sure you have`openslide-python` and `gunicorn` installed. And if the error still persists try to add path to the library specifically `libopenslide0` in my case it's this  `export LD_LIBRARY_PATH=\usr\share\doc\libopenslide0:$LD_LIBRARY_PATH`
I hope it helps

**Comment by We13b-MD** [comment]
Hello @barchakuz just saw this @gbengaoluwadahunsi has given you a correct answer y Oi can work with that and if any other issue raises feel free to.ket me know

**Reply by pycharm2003** [reply]
Hey @We13b-MD I am facing the issue @barchakuz faced. Here is my develop.yml file . Any changes that I made, are not being reflected in the UI on my localhost:/4010. Would be really helpful if you pinpoint the error. I have double checked the develop.yml file, stopped and removed ca-back, built it again, the issue still persists. 

version: '2'

services:
  mongo:
    image: mongo:4.2-bionic
    container_name: ca-mongo
    volumes:
      - ./db:/data/db
  back:
    build:
      context: "../Caracal"
      args:
        viewer: "develop"
        fork: "camicroscope"
    depends_on:
      - "mongo"
    ports:
      - "4010:4010"
    container_name: ca-back
    volumes:
      - ./config/login.html:/src/static/login.html
      - ./jwt_keys/:/src/keys/
      - ../caMicroscope:/src/camicroscope
      - ./config/routes.json:/src/routes.json
      - ./config/contentSecurityPolicy.json:/src/contentSecurityPolicy.json
    environment:
      JWK_URL: "https://www.googleapis.com/oauth2/v3/certs"
      IIP_PATH: "http://ca-iip:8080/fcgi-bin/iipsrv.fcgi"
      MONGO_URI: "mongodb://ca-mongo"
      DISABLE_SEC: "true"
      ALLOW_PUBLIC: "true"
  iip:
    build: "https://github.com/camicroscope/iipImage.git#develop"
    container_name: ca-iip
    restart: unless-stopped
    volumes:
      - ./images/:/images/
  loader:
    build: "https://github.com/camicroscope/SlideLoader.git#develop"
    container_name: ca-load
    volumes:
      - ./images/:/images/
      - ./cloud-upload-apis/:/cloud-upload-apis/
    environment:
      DICOM_PORT: "11112"
      DICOM_UI_PORT: "8042"
  dicomsrv:
    build: "https://github.com/camicroscope/dicomsrv.git"
    container_name: ca-dicomsrv
    restart: unless-stopped
    stdin_open: true
    tty: true
    ports:
      - "8042:8042"
      - "11112:11112"
    volumes:
      - ./jwt_keys/:/root/keys/
      - ./images/:/images/
      - ./config/OrthancConfiguration.json:/root/src/Configuration.json
    environment:
      DICOM_PORT: "11112"
      DICOM_UI_PORT: "8042"
      CARACAL_BACK_HOST_PORT: "ca-back:4010"

**Reply by barchakuz** [reply]
@pycharm2003 need to redo the process
docker-compose -f develop.yml build loader --no-cache
docker-compose -f caMicroscope.yml build
docker-compose -f caMicroscope.yml up
this will solve your issue

**Comment by SanjanaSogimatt** [comment]
Thanks, but after that I am not able build it successfully.
![error-3](https://github.com/camicroscope/caMicroscope/assets/87171143/74a2cfd0-e4b4-4618-8791-6a73e8cf292a)
This is what I am getting. Any help would be appreciated

**Reply by We13b-MD** [reply]
Hey @SanjanaSogimatt have you installed openslide library and if you have not go to the documentation 
2 clear cache and restart build process
Use ubuntu command line preferably 
Make sure you have a strong internet connection if the issue still arises let me know I will ready to help

**Reply by SanjanaSogimatt** [reply]
Yup installed all dependencies, I cleared the docker to build it again and I am getting the below error

**Comment by SanjanaSogimatt** [comment]
I basically did every solution which was on the internet but I'm still not able to build this and error I am getting is this
![error-4](https://github.com/camicroscope/caMicroscope/assets/87171143/31c1086a-c28e-4247-9aa9-5b85cf71b5b3)

**Reply by We13b-MD** [reply]
I had this issue also for days I found out that it was due to internet connection check your connection if the issue still arises I will have to take a closer look at it

**Reply by gbengaoluwadahunsi** [reply]
This might be a problem from the Caracal repo;

Check this thread, it is useful

https://github.com/orgs/camicroscope/discussions/758

**Comment by SanjanaSogimatt** [comment]
Guys, please anyone help

**Comment by gbengaoluwadahunsi** [comment]
Give me a minute to check the errors you are seeing from my GitHub account.

**Comment by gbengaoluwadahunsi** [comment]
Hello, I can see a error.png,

I guess you are trying to run this on a Linux machine.

Distro slide loader does not accept a png life, can you try a svs file?

I hope this helps.

**Comment by gbengaoluwadahunsi** [comment]
Hello, I can see an error.png,

I guess you are trying to run this on a Linux machine.

Distro slide loader does not accept a png life, can you try a svs file?

I hope this helps
